### ISSUE TITLE
const notDefined = notDefined; doesn't trigger no-undef

### ISSUE THREAD

**Issue by mhagmajer** [eslint/eslint#6363]
Eslint 2.11.1 doesn't show any errors (violations of rule `no-undef`) for:      

```js
const notDefined = notDefined;
```

**Comment by eslintbot** [comment]
Hi @mhagmajer, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
I can confirm that actually using `const notDefined = notDefined;` will throw in a browser, so it would be good if we could somehow catch this.

I'll try to help @mhagmajer out (please correct me if I missed something).

1. Version of ESLint rule: 2.11.1 (in OP)
1. What was done:
    * Source code: `const notDefined = notDefined;`
    * Config:  
    ```json
    {
        "rules": {
            "no-undef": "error"
        }
        "globals": {}
    }
    ```
1. Actual ESLint output: No error
1. `notDefined` (on the RHS of the assignment) should be flagged by no-undef

My guess is, we allow it for VariableDeclarations because `var` allows that pattern. But `let` and `const` do not allow it, and we should tighten the rule to fix it for those cases.

**Comment by mhagmajer** [comment]
Thanks Kevin!

**Comment by platinumazure** [comment]
@ilyavolodin Based on my analysis of the issue (which @mhagmajer seems to agree is correct), I think we might have a bug here-- unless we think that no-undef is not the place to catch this issue.

Personally I'm wondering if this should actually be handled by `no-use-before-define`.

@mhagmajer Do you have `no-use-before-define` enabled? If not, can you try it and let us know if that helps you with your issue?

**Comment by mhagmajer** [comment]
I'm pretty sure I do. We're using the Airbnb's eslint config.

**Comment by mhagmajer** [comment]
Either way, I believe that `no-undef` is the right rule to handle this to
prevent runtime errors.

**Comment by platinumazure** [comment]
I'm not sure I agree with that. `no-undef` is about flagging unintended use of global variables. As of the `const` declaration, the variable is no longer global, it is scoped (well, unless you're at top level scope and not using CommonJS or ES6 modules). `no-use-before-define` on the other hand seems like a much clearer case, because you're trying to use the variable before it is fully defined.

**Comment by mhagmajer** [comment]
Sure. What I meant is that we should include this logic in a rule that is
turned on by default so people don't have runtime errors on lint-correct
code.

**Comment by ilyavolodin** [comment]
@mhagmajer None of our rules are turned on by default. But `no-use-before-define` sounds like the right rule to put this in.

**Comment by mysticatea** [comment]
FYI, a related issue: #4280

**Comment by nzakas** [comment]
@mysticatea has this been solved already?

**Comment by mysticatea** [comment]
@nzakas No. This is a request to move the warning of `let a = a` case to `no-undef` rule from `no-use-before-define` rule. We have not moved it.

My position is neutral.
`no-use-before-define` rule is the proper rule to warn a use of variables before their definitions. On the other hand, I can understand this request. `no-use-before-define` rule is not perfect (it's the limitation of static analysis), so it would be helpful if `no-undef` warns clear TDZ violations.

**Comment by michaelficarra** [comment]
👍 Let's move it to `no-undef`.

**Comment by nzakas** [comment]
I'm not sure I agree with this suggestion. `no-undef` doesn't currently know anything about TDZ, it only warns when a variable hasn't been defined at all. `no-use-before-define` is explicitly for this case.  Plus, I think we cause more confusion by moving a check from one rule to another vs. just keeping it as-is.

**Comment by kaicataldo** [comment]
Doesn't seem like we've made any headway towards consensus on this in a while. Maybe time to close this?

**Comment by nzakas** [comment]
Agreed.
